### PR TITLE
[WIP] Zenuml support

### LINKED ISSUE #3435: Add ZenUML Support
Hey, author of ZenUML here.

**Is your feature request related to a problem? Please describe.**
ZenUML (ZenUML.com) is a diagram-as-code solution that convert text to sequence diagrams. ZenUML has a Confluence add-on, in which we integrated with Mermaid to extend the capability to more diagrams (many thanks). Many of our clients have asked us to unify the editing experience.

Screen Shot 2022-09-10 at 9 45 52 am

Screen Shot 2022-09-10 at 9 48 11 am

More introduction of ZenUML can be found here: https://modeling-languages.com/zenuml-drawing-sequence-diagram-easier-faster/

**Describe the solution you'd like**
I would like to integrate ZenUML into mermaid so that users can use ZenUML flavour DSL inside mermaid. ZenUML renderer is an open-source library (under MIT license). I am happy to implement the integration and pull-request.

**Describe alternatives you've considered**

- Confluence users use the ZenUML add-on - https://marketplace.atlassian.com/apps/1218380/zenuml-sequence-diagram?tab=overview&hosting=cloud
- JetBrains users use the ZenUML plugin - https://plugins.jetbrains.com/plugin/12437-zenuml-support
- Desktop users use our desktop applications - https://zenuml.com
- Some enterprise users integrate with our renderer library directly - https://www.npmjs.com/package/vue-sequence
- Most other users use the online editor - https://app.zenuml.com/ or https://zenuml.cn

Should you need more information, please let me know. Send me an email at peng(AT)zenuml.com or DM @PenngXiao on Twitter. Thanks.

### PR DESCRIPTION
**Pull request by MrCoder**
## :bookmark_tabs: Summary

Added support for ZenUML Diagram.

Resolves #3435 

## :straight_ruler: Design Decisions

ZenUML support code is in the `mermaid-zenuml` folder. Documents and Test are added to corresponding folders.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

File: packages/mermaid-zenuml/src/types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export {};
+
+declare global {
+  interface Window {
+    mermaid: any; // 👈️ turn off type checking
+  }
+}
```
Comment (sidharthv96) [inline packages/mermaid-zenuml/src/types/index.d.ts:7]: Is this necessary? 
`window.mermaid` should not be accessed as it most probably won't be present.

File: packages/mermaid-zenuml/src/zenumlRenderer.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  link.href = url;
+  root.getElementsByTagName('head')[0].appendChild(link);
+}
+
+// Create a Zen UML container outside of the svg first for rendering, otherwise the Zen UML diagram cannot be rendered properly
+function createTemporaryZenumlContainer() {
+  const container = document.createElement('div');
+  container.innerHTML = '<div id="app"></div>';
+  container.style.display = 'none';
+  const app = container.querySelector('#app') as HTMLElement;
```
Comment (sidharthv96) [inline packages/mermaid-zenuml/src/zenumlRenderer.ts:26]:
```suggestion
  container.innerHTML = '<div id="zenUMLApp"></div>';
  container.style.display = 'none';
  const app = container.querySelector('#zenUMLApp') as HTMLElement;
```

`app` seems too generic. How about `diagram` or `zenuml` ?

File: packages/mermaid-zenuml/src/zenumlRenderer.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  loadCss(root, './style.css');
+
+  const foreignObject = createForeignObject();
+  svgContainer.appendChild(foreignObject);
+
+  const { container, app } = createTemporaryZenumlContainer();
+  document.body.appendChild(container);
+
+  const zenuml = new ZenUml(app);
+  await zenuml.render(text, 'default');
```
Comment (sidharthv96) [inline packages/mermaid-zenuml/src/zenumlRenderer.ts:75]: What is `'default'`? Should it be configurable by the user?

File: .eslintrc.json [hunk 4]
```diff
@@ -54,7 +54,8 @@
       {
         "checkIdentifiers": false,
         "checkStrings": false,
-        "checkStringTemplates": false
+        "checkStringTemplates": false,
+        "customWordListFile": "./custom-word-list.txt"
```
Comment (sidharthv96) [inline .eslintrc.json:58]:
```suggestion
```

The word needs to be added in cspell.json. 
You can remove the custom-word-list.txt file.

File: packages/mermaid-zenuml/package.json [hunk 5]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "mermaid-zenuml",
+  "version": "9.2.2",
+  "description": "Markdownish syntax for generating flowcharts, sequence diagrams, class diagrams, gantt charts and git graphs.",
```
Comment (sidharthv96) [inline packages/mermaid-zenuml/package.json:4]: Update description to specify zenuml?

File: docs/syntax/zenuml.md [hunk 6]
```diff
@@ -0,0 +1,368 @@
+> **Warning**
+>
+> ## THIS IS AN AUTOGENERATED FILE. DO NOT EDIT.
+>
+> ## Please edit the corresponding file in [/packages/mermaid/src/docs/syntax/zenuml.md](../../packages/mermaid/src/docs/syntax/zenuml.md).
+
+# ZenUML
```
Comment (sidharthv96) [inline docs/syntax/zenuml.md:7]: Also needs to add zenuml in the sidebar so it's visible.

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
Looks good overall, but I wasn't able to actually get a diagram to render in docs. 

You can use the `pnpm docs:dev` in mermaid package to test.

Tried fixing the docs support, but wasn't able to push the changes. 
Adding it as a [patch](https://github.com/mermaid-js/mermaid/files/10129491/0001-Add-zenuml-in-docs.patch.txt)

```patch
From 9e206e0b1081ac08348e8ac243cee89121dfa71c Mon Sep 17 00:00:00 2001

---
 cSpell.json                                           | 3 ++-
 packages/mermaid/src/docs/.vitepress/theme/mermaid.ts | 5 +++--
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/cSpell.json b/cSpell.json
index 64187e1c..afb2cfde 100644
--- a/cSpell.json
+++ b/cSpell.json
@@ -87,7 +87,8 @@
     "visio",
     "vitepress",
     "xlink",
-    "yash"
+    "yash",
+    "zenuml"
   ],
   "patterns": [
     { "name": "Markdown links", "pattern": "\\((.*)\\)", "description": "" },
diff --git a/packages/mermaid/src/docs/.vitepress/theme/mermaid.ts b/packages/mermaid/src/docs/.vitepress/theme/mermaid.ts
index b287346f..59799fb2 100644
--- a/packages/mermaid/src/docs/.vitepress/theme/mermaid.ts
+++ b/packages/mermaid/src/docs/.vitepress/theme/mermaid.ts
@@ -1,8 +1,9 @@
 import mermaid, { type MermaidConfig } from 'mermaid';
-import mindmap from '@mermaid-js/mermaid-mindmap';
+import mindmap from '../../../../../mermaid-mindmap';
+import zenuml from '../../../../../mermaid-zenuml';
 
 try {
-  await mermaid.registerExternalDiagrams([mindmap]);
+  await mermaid.registerExternalDiagrams([mindmap, zenuml]);
 } catch (e) {
   console.error(e);
 }
-- 
2.32.1 (Apple Git-133)

```

But it seems like there is some error thrown. 

image

File: packages/mermaid-zenuml/src/zenumlRenderer.ts (hunk 3 above)
Comment (MrCoder) [reply packages/mermaid-zenuml/src/zenumlRenderer.ts:75]: This is a parameter for 'theme'. ZenUML supports a few built-in themes. This has not been included in this PR. Examples are: theme-black-white, theme-blue, theme-star-uml.
<div style="display: flex">
Screenshot 2022-12-01 at 4 45 47 pm
Screenshot 2022-12-01 at 4 45 34 pm
Screenshot 2022-12-01 at 4 45 23 pm
</div>

Comment (sidharthv96) [reply packages/mermaid-zenuml/src/zenumlRenderer.ts:75]: Ohh, ok.
We should open a new Issue to discuss the possibility of somehow mapping mermaid theme to this or letting the users configure this in mermaid config.

File: packages/mermaid-zenuml/src/types/index.d.ts (hunk 1 above)
Comment (MrCoder) [reply packages/mermaid-zenuml/src/types/index.d.ts]: I have removed this file.

File: packages/mermaid-zenuml/src/zenumlRenderer.ts (hunk 3 above)
Comment (MrCoder) [reply packages/mermaid-zenuml/src/zenumlRenderer.ts:60]: https://github.com/mermaid-js/mermaid/issues/3875 Issue has been created. Marking this as resolved.

File: docs/syntax/zenuml.md (hunk 6 above)
Comment (MrCoder) [reply docs/syntax/zenuml.md:7]: Done.

File: packages/mermaid-zenuml/package.json (hunk 5 above)
Comment (MrCoder) [reply packages/mermaid-zenuml/package.json]: Updated

File: .vite/build.ts [hunk 7]
```diff
@@ -41,11 +41,16 @@ const packageOptions = {
...
+  'mermaid-zenuml': {
+    name: 'mermaid-zenuml',
+    packageName: 'mermaid-zenuml',
+    file: 'detector.ts',
+  },
+  'mermaid-example-diagram': {
+    name: 'mermaid-example-diagram',
+    packageName: 'mermaid-example-diagram',
+    file: 'diagram-definition.ts',
+  },
```
Comment (sidharthv96) [inline .vite/build.ts:53]: Please comment this out. We want to avoid wasting time during build.

File: packages/mermaid-zenuml/README.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  // ...
+  const store = VueSequence.Store();
+  store.state.code = text;
+
+  new Vue({
+    el: diagram._groups[0][0].parentNode,
+    store,
+    render: (h) => h(VueSequence.DiagramFrame),
+  });
+};
```
Comment (sidharthv96) [inline packages/mermaid-zenuml/README.md:104]: Is this outdated now?

File: packages/mermaid-zenuml/package.json [hunk 9]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "mermaid-zenuml",
+  "version": "9.2.2",
+  "description": "ZenUML is a text to sequence diagram renderer. For more information, please checkout https://ZenUML.com",
```
Comment (sidharthv96) [inline packages/mermaid-zenuml/package.json:4]:
```suggestion
  "description": "Mermaid integration for ZenUML, a text to sequence diagram renderer. For more information, please checkout https://ZenUML.com",
```

**Review by sidharthv96 — APPROVED** [review]
Awesome! LGTM :)

**Comment by sidharthv96** [comment]
Unfortunately, there appears to be some subtle bugs. 

The diagrams load fine when I navigate to zenuml page by clicking on sidebar link.

image

But the diagrams disappear when:
- Switching the theme using the button at the top.
- Relaoding the page

image

It's not readable when darkmode is enabled. 
image

**Comment by MrCoder** [comment]
I'll have a look. Thanks for your testing.

**Comment by MrCoder** [comment]
> Unfortunately, there appears to be some subtle bugs.
> 
> The diagrams load fine when I navigate to zenuml page by clicking on sidebar link.

It seems that mindmap on the `develop` branch has the same issue. I was able to 'fix' it by removing `id` from svg. But I think there should be better approaches.

```javascript
// Mermaid.vue
  console.log({ mermaidConfig });
  let value = await render(props.id, decodeURIComponent(props.graph), mermaidConfig);
  // remove id from svg
  value = value.replace(/id="mermaid-[0-9]+"/, '');
  svg.value = value;

```

**Comment by MrCoder** [comment]
This PR is replaced by https://github.com/mermaid-js/mermaid/pull/3983. The new PR include ZenUML/core source code. The new PR will give a much cleaner commit history.

**Pull request closed without merging**